### ISSUE TITLE
`object not found` when trying to pull a repository cloned with `Depth: 1`

### ISSUE THREAD

**Issue by stingalleman** [go-git/go-git#305]
Hi,

When you clone a repository with `Depth: 1`, you cannot pull it. It'll error with `object not found`.

Cloning:
```go
_, err := git.PlainClone(repoPath, false, &git.CloneOptions{
	URL: gitURL,
	Depth: 1,
	})
```

Pulling:
```go
// Open repo
r, err := git.PlainOpen(repo.Path)
if err != nil {
	panic(fmt.Errorf("error while opening git repo (%s) %s", repo.Name, err))
}
// Get worktree
tree, err := r.Worktree()
if err != nil {
	panic(fmt.Errorf("error while opening git repo (%s) %s", repo.Name, err))
}

// Pull
err = tree.Pull(&git.PullOptions{})
// If repo is already up to date, do nothing
if err == git.NoErrAlreadyUpToDate {
	// do nothing
} else if err != nil {
	panic(fmt.Errorf("error while pulling git repo (%s) %s", repo.Name, err))
}
```

It's critical that I can pull with Depth: 1, because the repository I'm pulling might be very large (hundreds of thousands of commits). Can somebody take a look at this?

**Comment by stingalleman** [comment]
@jfontan @mcarmonaa any idea?

**Comment by wyarde** [comment]
TL;DR - Use `file://path/to/repo` and you should be good

I ran into similar problems recently. Initially I worked around the shallow fetch issue by doing a full fetch. The fetch was actually fast, but then the push itself was taking 10 seconds. Weird... So I ended up spending more time trying to understand what was going wrong. 

Unfortunately I'm not familiar enough with the code base to fix the actual issue, but hopefully my brain dump helps for people to locate the issue, or at least provides useful clues how to work around it:

Without further ado, here are my findings:

For local urls, the code in `revlist.go // ObjectsWithStorageForIgnores` constructs a list of all objects to be pushed. Internally, it constructs a list of objects to be ignored first. This ignore list also includes the commits which as a result of the shallow fetch don't have a parent present:
```
	ignore, err := objects(ignoreStore, ignore, nil, true)
	if err != nil {
		return nil, err
	}

	return objects(s, objs, ignore, false)
```

The first issue I ran into when doing a shallow fetch and then a push is I think what you run into as well. The first `objects()` call is unable to locate the objects to be ignored, because the path to the objects it constructed did not include the `.git` folder. That's where I realized I provided the local repository path without ".git". Formally this I guess is a mistake on my side since documentation I could find on how to clone a local repository says you should point to the ".git" folder. However, the git CLI doesn't care it works in either case. And also go-git seems to largely work and just fail in this unexpected way. So I think an improvement here would be to either enforce the path in the expected format, or fix the logic to work either way.

However when including ".git" in the path url, I ran into the next issue: The first call to `objects()` in the code above, the one to obtain all objects to ignore, was taking 5 seconds to complete. It returned a list of 32K objects which I presume is the full repository. Well that's twice as fast as the non-shallow fetch/push, but still horribly slow to push just a single object. 

So then I realized I could bypass the "local url" logic by providing the repository in the `file://path/to/repo` format (Including `.git` is no longer required). The push now takes less than .1 second.

**Comment by stingalleman** [comment]
@wyarde thank you so much for posting your findings! I’m going to try this when I’m home- thank you so much

**Comment by stingalleman** [comment]
I hope that the maintainers can fix this bug with this information.

**Comment by stingalleman** [comment]
Hey @wyarde, I tried opening the repository with `file://`, but it can't find the actual repository now. 
```go
r, err := git.PlainOpen(repo.Path)
```

**Comment by wyarde** [comment]
What does your file path look like?
If this is on windows, you'll need to use forward slashes, for example `file://c:/my/path/to/the/repo`.

**Comment by wyarde** [comment]
(If not on windows, you would still need to use forward slashes, but I guess it's more intuitive then)

**Comment by marguerite** [comment]
same issue here, the code I use is exactly the plain clone code the issue author mentioned except for the gitURL which is “https://github.com/rime/rime-cantonese”

and here’s the things I found so far.

the commit treeobject got from getTreeFromCommitHash in worktree.Reset() is still good, you can print the names etc. but when getting the actual file with t.File in checkoutChangeRegularFile, the file blob can’t have a Reader(). eg in plumbing/object/object.go. the blob obj is actually a plumbing.EncodedObject.

And then I lost my direction because EncodedObject is an interface, there are so many ways in go-git to create one. 

BTW, the file in trouble in my previous git url is “jyut6ping3.dict.yaml” and the commit in trouble is “ 1168d4ca475656748938475ca5d96e41096d8b5a“.

Hope this helps for who interested in further debugging. For now I have to add a “use-system-git” option for my app...sad

**Comment by marguerite** [comment]
aha...sorry for my example, you can’t even plain clone https://github.com/rime/rime-cantonese

**Comment by stingalleman** [comment]
@jfontan @mcarmonaa any idea?

**Comment by stingalleman** [comment]
> What does your file path look like?
> If this is on windows, you'll need to use forward slashes, for example `file://c:/my/path/to/the/repo`.

So sorry @wyarde I completely forgot and have been really, really busy. My file path is just `file://Users/stingalleman/git-repo`

**Comment by wyarde** [comment]
@stingalleman is this on windows? You need to include the drive letter as in my example

**Comment by stingalleman** [comment]
@wyarde No, I am on a MacOS system.

**Comment by wyarde** [comment]
@stingalleman sorry cannot help you there

**Comment by stingalleman** [comment]
Hi @jfontan @mcuadros, can ypou take a look at this? Still waiting :-)

**Comment by laszlocph** [comment]
I'm getting `empty git-upload-pack given` for the same scenario, with `v5.4.2`

**Comment by rotty3000** [comment]
A colleague and I have solved this issue. The iterator used in the fastforward check wasn't prepared for the case when a commit might have a missing parent (as in shallow clones). Handling that case solved the problem of pulling on shallow clones.

**Comment by rotty3000** [comment]
I think I've actually found a better solution to this problem. It seems that the concept of `haves` v.s `wants` is not balanced in that shallowness is only checked in the `wants` side of the checks but not in the `haves` side.

This iteration should check if `c.Hash` is a shallow reference, something like this:

```patch
@@ -691,6 +691,15 @@ func getHavesFromRef(
 	toVisit := maxHavesToVisitPerRef
 	return walker.ForEach(func(c *object.Commit) error {
 		haves[c.Hash] = true
+
+		if s, _ := s.Shallow(); len(s) > 0 {
+			for _, sh := range s {
+				if sh == c.Hash {
+					return storer.ErrStop
+				}
+			}
+		}
+
 		toVisit--
 		// If toVisit starts out at 0 (indicating there is no
 		// max), then it will be negative here and we won't stop
```

**Comment by williambanfield** [comment]
Hey! I was wondering when we may expect this to be released? I have a few changes that are ready to make use of this change and would be interested to know the timeline.
Thanks!

**Comment by pjbgf** [comment]
@williambanfield I am not sure when the next release will happen, but it will probably be by the end of this month. If you need the changes sooner, use the pseudo-version instead:

`go get github.com/go-git/go-git/v5@master`

**Comment by jmriebold** [comment]
Is this fix actually working for anyone? Because even with v5.8.1 I'm getting the same `ErrObjectNotFound` any time a commit is pushed to a branch in a shallow-cloned repo.

**Comment by AriehSchneier** [comment]
@jmriebold this bug was about pulling/fetching, if you are having issues with pushing then please see if there is another issue with that bug or open a new one.

**Comment by jmriebold** [comment]
Sorry, what I mean is that I still get `object not found` when I try to pull a new commit on a shallow-cloned repo.

**Comment by jmriebold** [comment]
@AriehSchneier do you have any suggestions? If not I think this issue should be reopened.

**Comment by AriehSchneier** [comment]
@jmriebold are you able to write a test, or tell us the steps to be able to reproduce it?

**Comment by stvnksslr** [comment]
@AriehSchneier ill put together something that can just be checkedout and run but i can reproduce with the following.

**Comment by AriehSchneier** [comment]
Ohh, you are trying to change the depth (unshallow), that is probably being tracked in #328

**Comment by stvnksslr** [comment]
Thank you for responding! I really appreciate you helping us get to the root of this issue.

To be clear no this is not unshallowing related, its a cloned repo with a depth of 1 and a pull with a depth of 1 returning "object not found" when a change is made in the repo and the pull is initiated.

**Comment by stvnksslr** [comment]
@pjbgf @AriehSchneier Apologize for the tag, the steps for reproducing this issue seem straight forward unsure as to why it wouldn't show up in the tests. let me know if there is anything else I can do to help with any of the leg work for this issue.

Additionally I can confirm the issue is not present with the same code example if the depth flags are removed from both the clone and the pull.

**Comment by Tang8330** [comment]
Got really excited about https://github.com/go-git/go-git/pull/932, updated dependency to use this...but still encountering the git pull problem.

Any workarounds and also ideas on where this is happening?

**Comment by dhoizner** [comment]
> Got really excited about #932, updated dependency to use this...but still encountering the git pull problem.
> 
> Any workarounds and also ideas on where this is happening?

+1 from me- the behavior that i'm running into is:
- pulls on a shallow clone when there have been no updates to the remote branch succeed
- pulls on a shallow clone when there are new changes in the remote branch faIl with the "object not found" error

**Comment by dvachaiev-hdx** [comment]
Seems the issue is present in v6 again.

**Comment by aljoshare** [comment]
> Seems the issue is present in v6 again.

+1, I can't iterate the log if depth is 1. What works:

- Clone with depth > 1
- Clone without depth, even if the repo has only one commit

Should we track that in a new issue?

**Comment by pjbgf** [comment]
Re-opening as per feedback above.

**Comment by hieroglyphiks** [comment]
How were you able to repro the issue? I wasn't able to get it to reoccur with a simple function using `@main`?

```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/go-git/go-git/v6"
	"github.com/go-git/go-git/v6/plumbing/object"
)

func main() {
	originalRepoPath := "./original-repo"
	clonePath := "./cloned-repo"

	// Clean up on exit
	defer func() {
		fmt.Println("\nCleaning up...")
		os.RemoveAll(originalRepoPath)
		os.RemoveAll(clonePath)
	}()

	// Remove existing directories if they exist
	os.RemoveAll(originalRepoPath)
	os.RemoveAll(clonePath)

	// Step 1: Create a local git repository
	fmt.Println("Creating original repository...")
	originalRepo, err := git.PlainInit(originalRepoPath, false)
	if err != nil {
		fmt.Printf("Error creating repository: %v\n", err)
		os.Exit(1)
	}

	// Create an initial commit in the original repo
	w, err := originalRepo.Worktree()
	if err != nil {
		fmt.Printf("Error getting worktree: %v\n", err)
		os.Exit(1)
	}

	// Create a file
	filename := originalRepoPath + "/README.md"
	err = os.WriteFile(filename, []byte("# Test Repository\n"), 0644)
	if err != nil {
		fmt.Printf("Error creating file: %v\n", err)
		os.Exit(1)
	}

	// Add and commit
	_, err = w.Add("README.md")
	if err != nil {
		fmt.Printf("Error adding file: %v\n", err)
		os.Exit(1)
	}

	_, err = w.Commit("Initial commit", &git.CommitOptions{
		Author: &object.Signature{
			Name:  "Test User",
			Email: "test@example.com",
			When:  time.Now(),
		},
	})
	if err != nil {
		fmt.Printf("Error committing: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("Original repository created with initial commit")

	// Step 2: Plain clone it with depth 1
	fmt.Printf("\nCloning from %s with depth 1...\n", originalRepoPath)
	clonedRepo, err := git.PlainClone(clonePath, &git.CloneOptions{
		URL:   originalRepoPath,
		Depth: 1,
	})
	if err != nil {
		fmt.Printf("Error cloning: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("Clone successful!")

	// Step 3: Verify that the remote of the cloned repo points to the original
	fmt.Println("\nVerifying remote configuration...")
	remotes, err := clonedRepo.Remotes()
	if err != nil {
		fmt.Printf("Error getting remotes: %v\n", err)
		os.Exit(1)
	}

	found := false
	for _, remote := range remotes {
		if remote.Config().Name == "origin" {
			urls := remote.Config().URLs
			if len(urls) > 0 && urls[0] == originalRepoPath {
				fmt.Printf("✓ Remote 'origin' correctly points to: %s\n", originalRepoPath)
				found = true
			}
		}
	}
	if !found {
		fmt.Println("✗ Remote 'origin' does not point to original repository")
		os.Exit(1)
	}

	// Step 4: Make a commit to the original repository
	fmt.Println("\nMaking a new commit to original repository...")
	err = os.WriteFile(filename, []byte("# Test Repository\n\nUpdated content\n"), 0644)
	if err != nil {
		fmt.Printf("Error updating file: %v\n", err)
		os.Exit(1)
	}

	_, err = w.Add("README.md")
	if err != nil {
		fmt.Printf("Error adding file: %v\n", err)
		os.Exit(1)
	}

	_, err = w.Commit("Second commit", &git.CommitOptions{
		Author: &object.Signature{
			Name:  "Test User",
			Email: "test@example.com",
			When:  time.Now(),
		},
	})
	if err != nil {
		fmt.Printf("Error committing: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("New commit created in original repository")

	// Step 5: Attempt to pull in the cloned repo
	fmt.Println("\nAttempting to pull in cloned repository...")
	clonedWorktree, err := clonedRepo.Worktree()
	if err != nil {
		fmt.Printf("Error getting cloned worktree: %v\n", err)
		os.Exit(1)
	}

	err = clonedWorktree.Pull(&git.PullOptions{
		RemoteName: "origin",
		Depth: 1,
	})

	if err == git.NoErrAlreadyUpToDate {
		fmt.Println("Pull result: Already up-to-date")
	} else if err != nil {
		fmt.Printf("Pull result: Error - %v\n", err)
	} else {
		fmt.Println("Pull result: ✓ Successfully pulled new changes")
	}
}
```

gets me 
```
$ go run main.go
Creating original repository...
Original repository created with initial commit

Cloning from ./original-repo with depth 1...
Clone successful!

Verifying remote configuration...
✓ Remote 'origin' correctly points to: ./original-repo

Making a new commit to original repository...
New commit created in original repository

Attempting to pull in cloned repository...
Pull result: ✓ Successfully pulled new changes

Cleaning up...
```
